### PR TITLE
Day 4 commit

### PR DESCRIPTION
**Pull request by raghavsrm**

### CODE REVIEW

File: src/com/raghav/javabasics/day4/CountIntCharSpecialCharInString.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+//Write a program to count number of integers, number of characters and number of special characters.
+public class CountIntCharSpecialCharInString {
+    public static void main(String[] args) {
+        CountIntCharSpecialCharInString result = new CountIntCharSpecialCharInString();
+        result.charCount("abcdefghijklmnopqrstuvwxyz0123456789 !@#$%^&*()_+");
+        result.charCount("");
+        result.charCount(null);
+        result.charCount("cccc0000c ");
+        result.charCount("ascisa%%%-1");
+        result.charCount("abhddabc123*(&");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/CountIntCharSpecialCharInString.java:11]: Output:
Letter count: 26   Number Count: 10   Special Char Count: 13
Array is Empty/Null
Array is Empty/Null
Letter count: 5   Number Count: 4   Special Char Count: 1
Letter count: 6   Number Count: 1   Special Char Count: 4
Letter count: 8   Number Count: 3   Special Char Count: 3

File: src/com/raghav/javabasics/day4/CountOccurrencesOfCharInString.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.raghav.javabasics.day4;
+//Write a java program to count occurrences of each character in String in java.
+
+public class CountOccurrencesOfCharInString {
+    public static void main(String[] args) {
+        CountOccurrencesOfCharInString result = new CountOccurrencesOfCharInString();
+        result.charCount("abc");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/CountOccurrencesOfCharInString.java:7]: Output: 
a: 1  b: 1  c: 1  
Array is Empty/Null
Array is Empty/Null
0: 4  c: 5  
#: 1  $: 3  %: 2  &: 2  (: 1  ^: 1  a: 2  c: 1  s: 1  
 : 8  a: 2  b: 2  c: 1  d: 2  h: 1

File: src/com/raghav/javabasics/day4/FindDuplicatesInArray.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
+package com.raghav.javabasics.day4;
+
+// Given an array of numbers, write a program to find the duplicates.
+
+public class FindDuplicatesInArray {
+    boolean abc= false;
+    public static void main(String[] args) {
+        FindDuplicatesInArray result = new FindDuplicatesInArray();
+        result.findDuplicates(null);
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/FindDuplicatesInArray.java:9]: Output:
Array is Null
Array is Null
 23  -5  1  0 
 1  2  55  0 
 0

File: src/com/raghav/javabasics/day4/FindFactorialOfNumber.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package com.raghav.javabasics.day4;
+//Write a java program to find factorial of a given number.
+
+public class FindFactorialOfNumber {
+    public  static void main(String args[]){
+        FindFactorialOfNumber result = new FindFactorialOfNumber();
+        result.findFactorial(40);
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/FindFactorialOfNumber.java:7]: output:
Enter integers between 1 - 40
Enter integers between 1 - 40
3628800
6402373705728000
Enter integers between 1 - 40
1

File: src/com/raghav/javabasics/day4/FindFirstNonRepeatingChar.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package com.raghav.javabasics.day4;
+//Given a string, write a program to print the first non-repeating character in the String.
+
+public class FindFirstNonRepeatingChar {
+    public static void main(String args[]) {
+        FindFirstNonRepeatingChar result = new FindFirstNonRepeatingChar();
+        result.findNonRepeatingChar("abhddabc");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/FindFirstNonRepeatingChar.java:7]: Output:
h
c
c
h
a
Array is Empty/Null
Array is Empty/Null

File: src/com/raghav/javabasics/day4/FindFirstNonRepeatingCharAscii.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package com.raghav.javabasics.day4;
+//Given a string, write a program to print the first non-repeating character in the String.
+
+public class FindFirstNonRepeatingCharAscii {
+    public static void main(String args[]) {
+        FindFirstNonRepeatingCharAscii result = new FindFirstNonRepeatingCharAscii();
+        result.findNonRepeatingChar("abc");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/FindFirstNonRepeatingCharAscii.java:7]: Output: 
a
c
No non repetitive num in string
c
h
Array is Empty/Null
Array is Empty/Null

File: src/com/raghav/javabasics/day4/FindHighestOccurredCharInString.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package com.raghav.javabasics.day4;
+// Given a string, write a program to return the highest occurred character in a String.
+public class FindHighestOccurredCharInString {
+    public static void main(String[] args) {
+        FindHighestOccurredCharInString result = new FindHighestOccurredCharInString();
+        result.charCount("abc");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/FindHighestOccurredCharInString.java:6]: Output:
MAX occurred char is: a - 1, MAX occurred char is: b - 1, MAX occurred char is: c - 1, 
Array is Empty/Null
MAX occurred char is: c - 5, 
MAX occurred char is: $ - 3, MAX occurred char is: 1 - 3, MAX occurred char is: 2 - 3, MAX occurred char is: 3 - 3, 
MAX occurred char is: a - 2, MAX occurred char is: b - 2, MAX occurred char is: d - 2, 
Array is Empty/Null

File: src/com/raghav/javabasics/day4/FindSquareRootOfNumber.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package com.raghav.javabasics.day4;
+
+public class FindSquareRootOfNumber {
+    public static void main(String args[]) {
+        FindSquareRootOfNumber result = new FindSquareRootOfNumber();
+        result.findSqrRoot(25);
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/FindSquareRootOfNumber.java:6]: Output :  only gives accurate result for perfect square integers.
5
3

File: src/com/raghav/javabasics/day4/PrintDuplicateCharInString.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package com.raghav.javabasics.day4;
+//Given a string, write a program to print the duplicate characters in the String.
+
+public class PrintDuplicateCharInString {
+    boolean abc= false;
+    public static void main(String[] args) {
+        PrintDuplicateCharInString result = new PrintDuplicateCharInString();
+        result.findDuplicates("program to print the duplicate characters in the String.");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/PrintDuplicateCharInString.java:8]: Output:
[p][r][o][g][a][ ][t][i][n][h][e][c][ ]Array is Empty/Null
Array is Empty/Null
[a][b][ ]

File: src/com/raghav/javabasics/day4/RemoveWhiteSpacesInString.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
+package com.raghav.javabasics.day4;
+//Write a java program to remove all the white-spaces in the String.
+
+public class RemoveWhiteSpacesInString {
+    public static void main(String args[]){
+        RemoveWhiteSpacesInString result = new RemoveWhiteSpacesInString();
+        result.removeSpaces("Write a java program, to remove all the white-spaces in the String");
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/RemoveWhiteSpacesInString.java:7]: Output:
 W r i t e a j a v a p r o g r a m , t o r e m o v e a l l t h e w h i t e - s p a c e s i n t h e S t r i n g
Array is Empty/Null
Array is Empty/Null

File: src/com/raghav/javabasics/day4/SplitStringIntoSubstrings.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
+package com.raghav.javabasics.day4;
+// Given a string and an index, write a program to split the string into two substrings based on the index.
+
+public class SplitStringIntoSubstrings {
+    public static void main(String[] args) {
+        SplitStringIntoSubstrings result = new SplitStringIntoSubstrings();
+        result.splitString("",21);
```
Comment (raghavsrm) [inline src/com/raghav/javabasics/day4/SplitStringIntoSubstrings.java:7]: Output:
String is Empty/Null
String is Empty/Null
First substring: a   Second substring: bcdefg
Enter Valid Index
First substring: abcde   Second substring: fg
Enter Valid Index
First substring: abcdefg   Second substring: 
First substring: abc      Second substring: defg
First substring: ab$^%.   Second substring: .cdefg
Enter Valid Index

File: src/com/raghav/javabasics/day4/CountIntCharSpecialCharInString.java (hunk 1 above)
Comment (mspranav) [reply src/com/raghav/javabasics/day4/CountIntCharSpecialCharInString.java:11]:
```# of int, char and spe char

Input = “”
for(int i=0;i<=input.length;i++){
	If input[i].isAlpha() {
		letterCount += 1
}
	If input[i].isDigit() {
		digitCount += 1
}
If input[i].isSpecialChar() {
		specialCount += 1
}

}
```

File: src/com/raghav/javabasics/day4/CountOccurrencesOfCharInString.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        result.charCount("");
+        result.charCount(null);
+        result.charCount("cccc0000c");
+        result.charCount("asc^%$(#&%$&$a");
+        result.charCount("abh        ddabc");
+        }
+
+        public void charCount(String str) {
+            System.out.println();
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/CountOccurrencesOfCharInString.java:16]: Not Needed

File: src/com/raghav/javabasics/day4/CountIntCharSpecialCharInString.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        int specialCount=0;
+
+        for(int i=0; i< arr.length;i++){
+            if((i>=65 && i<=90)|| (i>=97 && i<= 122)){
+                if(arr[i]==1){
+                    letterCount+=1;
+                }
+                else if(arr[i]>1){
+                    letterCount=letterCount+arr[i];
+                }
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/CountIntCharSpecialCharInString.java:39]: Cheange as Discussed to the above logic

File: src/com/raghav/javabasics/day4/FindFirstNonRepeatingCharAscii.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+        for (int j = 0; j < str.length(); j++) {
+            val = (int) str.charAt(j);
+            if (arr[val] == 1) {
+                System.out.println(str.charAt(j));
+                break;
+            }
+            else {
+                if(j==str.length()-1){
+                    System.out.println("No non repetitive num in string");
+                }
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/FindFirstNonRepeatingCharAscii.java:37]: Not Needed, remove.
Put it after the loop.

File: src/com/raghav/javabasics/day4/FindHighestOccurredCharInString.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+            System.out.print("Array is Empty/Null");
+            return;
+        }
+        int[] myArray = new int[128];
+        int val;
+        for (int i = 0; i < str.length(); i++) {
+            val = (int) str.charAt(i);
+            myArray[val] += 1;
+        }
+            int swapValue1 = 1; int swapValue2= 0;
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/FindHighestOccurredCharInString.java:27]: Nit: Unused Variable and Please do add/use meaningful names to variables

File: src/com/raghav/javabasics/day4/FindSquareRootOfNumber.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    public void findSqrRoot(int x) {
+        if (x == 0 || x == 1) {
+            System.out.println(x);
+        }
+        int i = 1, result = 1;
+        while (result <= x){
+            i++;
+            result = i * i;
+        }
+        System.out.println(i - 1);
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/FindSquareRootOfNumber.java:20]: extra loop, not needed. Explained

File: src/com/raghav/javabasics/day4/PrintDuplicateCharInString.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+            return;
+        }
+        char[] newArray = new char[str.length()];
+        int count = 0;
+        for (int i = 0; i < str.length(); i++) {
+            for (int j = i+1; j < str.length(); j++) {
+                if (str.charAt(i) == str.charAt(j) && searchNewArray(newArray, str.charAt(i)) == true) {
+                    newArray[count] = str.charAt(i);
+                    count += 1;
+                    abc=true;
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/PrintDuplicateCharInString.java:25]: Meaningful Variables (flag maybe?)

**Review by mspranav — CHANGES_REQUESTED** [review]
Please Go through the comments and make the required changes

File: src/com/raghav/javabasics/day4/FindFirstNonRepeatingCharAscii.java [hunk 18]
```diff
@@ -25,18 +25,14 @@ public void findNonRepeatingChar(String str) {
             val = (int) str.charAt(i);
             arr[val] += 1;
         }
-        for (int j = 0; j < str.length(); j++) {
+        int j=0;
+        for (j = 0; j < str.length(); j++) {
```
Comment (mspranav) [inline src/com/raghav/javabasics/day4/FindFirstNonRepeatingCharAscii.java:29]: Unnecessary initialization, can be initialized within the loop conditions

**Review by mspranav — APPROVED** [review]

**Pull request merged**